source/common/event/dispatcher_impl.cc

```diff
@@ -121,6 +126,20 @@ Network::DnsResolverSharedPtr DispatcherImpl::createDnsResolver(
     const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
     const bool use_tcp_for_dns_lookups) {
   ASSERT(isThreadSafe());
+#ifdef __APPLE__
+  static bool use_apple_api_for_dns_lookups =
+      Runtime::runtimeFeatureEnabled("envoy.restart_features.use_apple_api_for_dns_lookups");
+  if (use_apple_api_for_dns_lookups) {
+    RELEASE_ASSERT(
```

Comment:
Shouldn't we be rejecting the config earlier in a more graceful way? If so, shouldn't these be `ASSERT`s?

Reply:
Yes, this should be exceptions that fail during config load.

Reply:
It all comes down to snapping the runtime once by making it static local state here vs checking earlier in callers of this function. 

I guess I could change this function to return nullptr on failure, and have callers deal with the return value?

Reply:
I believe the DNS resolver is created during config load time in all cases. I think you can just throw an exception here.

Reply:
I think that's the case for all xDS, assuming this also holds for dynamic forward proxy, SG.

Reply:
It holds for the dynamic proxy. Let me check in the UDP filter. If that's the case, I'll update with an exception.

Reply:
@htuch @mattklein123 unfortunately, the UDP DNS filter creates a resolver (underneath its own wrapper) for every filter instance construction: 

https://github.com/envoyproxy/envoy/blob/3b5acb2f43548862dadb243de7cf3994986a8e04/source/extensions/filters/udp/dns_filter/dns_filter.cc#L229

So can't throw an exception as is (and per @htuch the assert is not great either). 

The underlying resolver in the UDP DNS filter should _definitely_ be shared, and I can work to untangle. However, I don't want to block this PR too much (we need it to be able to run envoy mobile on ios 14 devices). How do you feel about landing this PR with a warn log in place of the assert. Then I can work on untangling the filter situation, and subsequently upgrading the warn log here into an exception?

Reply:
I would actually rather you just leave the RELEASE_ASSERT and make sure the error message is good so the OSX user can slip the runtime flag, then come back and clean it up. So just leave a TODO for that? @htuch?